SYCL/KernelAndProgram/spec_constants_after_link.cpp

```diff
@@ -0,0 +1,106 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -fsycl-device-code-split=per_kernel %s -o %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// FIXME: ACC devices use emulation path, which is not yet supported
+
+// This test checks that specialization constant information is available on
+// kernel bundles produced by sycl::link.
+
+#include <sycl/sycl.hpp>
+
+#include <optional>
+#include <vector>
+
+using namespace sycl;
+
+class Kernel1;
+class Kernel2;
+
+struct TestStruct {
+  float f;
+  long long ll;
+};
+
+bool operator==(const TestStruct &LHS, const TestStruct &RHS) {
+  return LHS.f == RHS.f && LHS.ll == RHS.ll;
+}
+bool operator!=(const TestStruct &LHS, const TestStruct &RHS) {
+  return !(LHS == RHS);
+}
+
+constexpr specialization_id<TestStruct> SpecName1;
+constexpr specialization_id<int> SpecName2;
+
+void UnusuedFunc(queue &Q) {
+  Q.single_task<Kernel1>([=](kernel_handler KHR) {
+    volatile TestStruct A = KHR.get_specialization_constant<SpecName1>();
+  });
+
+  Q.single_task<Kernel2>([=](kernel_handler KHR) {
+    volatile int A = KHR.get_specialization_constant<SpecName2>();
+  });
+}
```

Comment:
I'd prefer if we execute the code on device and communicate the value of the constant back to the host to verify it.

Reply:
We can do that, but that is outside the intention of the test. The intention is just to make sure we can answer the host-side queries. Because it is done using kernel bundles to set the specialization constants and link it, it would become a much more complicated test if we wanted to run the kernels too, so seeing as it would be testing something unrelated I am reluctant to change it.

Reply:
See https://github.com/intel/llvm-test-suite/pull/1560#issuecomment-1410888309.